### PR TITLE
feat: asdf latest defer to plugin to determine the latest version

### PR DESCRIPTION
**Pull request by threkk**
# Summary

If the command `latest-stable` exists in the plugin, `asdf latest` defers to it to determine the latest stable version.

Fixes: #763 
Resolution of #648 now falls to plugin authors.

### CODE REVIEW

**Comment by threkk** [comment]
I have opened this pull request which solves an existing an existing issue about three weeks ago, but it has gotten no attention since then. Is there anything missing or anything I can do so it can be taken in consideration? :(

**Comment by jthegedus** [comment]
@threkk Sorry mate, will look at this in the next couple of days. We appreciate the help and this PR isn't going unnoticed. Sometimes I don't look at a PR until I know I have enough time to give it attention. Which these past few weeks hasn't been much time at all.

Will look at this next.

**Comment by jthegedus** [comment]
@Stratus3D I am confident the changes here so will push through.

I will clean up the testing in a future PR as we need to test most of the code against 2 different plugins, ones with the minimum scripts and another with all optional scripts like `latest-stable` as added here. In the meantime I have used `dummy_legacy_plugin` though this is incorrect.

Additionally, I will PR some docs updates about plugin API. We're currently handling calling out to scripts from plugins in different ways, expecting different results, eg: `list-all` does a lot more work to capture plugin reported errors, whereas here (`latest-stable`) we expect something or nothing and no `stderr`s, though we're not guarding against plugin reported errors in these scripts.

**Review by jthegedus — APPROVED** [review]
@threkk Thanks for contributing this. I added more testing scenarios and merged with the changes on `master` that came from #633

**Comment by threkk** [comment]
Thank you very much for taking it in consideration. This will also allow me to finish a plug-in I was working on too :)

File: lib/commands/command-latest.bash [hunk 1]
```diff
@@ -5,18 +5,31 @@ latest_command() {
...
-  versions=$(asdf list-all "$plugin_name" "$query" |
-    grep -vE "(^Available versions:|-src|-dev|-latest|-stm|[-\\.]rc|-alpha|-beta|[-\\.]pre|-next|(a|b|c)[0-9]+|snapshot|master)" |
-    sed 's/^\s\+//' |
-    tail -1)
-
-  if [ -z "${versions}" ]; then
-    exit 1
+
+  if [ -f "${plugin_path}/bin/latest-stable" ]; then
+    versions=$(bash "${plugin_path}"/bin/latest-stable "$query")
```
Comment (Stratus3D) [inline lib/commands/command-latest.bash:18]: FYI at some point hopefully in the near future I'd like to explore allowing callbacks to be written in any language and not just bash. We'd need to change the callback invocations from `bash "${plugin_path}"/bin/<callback> <args>` to just `"${plugin_path}"/bin/<callback> <args>`. This will give plugin creators more control over how they implement callbacks.

File: lib/commands/command-latest.bash [hunk 2]
```diff
@@ -5,18 +5,31 @@ latest_command() {
...
+    if [ -z "${versions}" ]; then
+      # this branch requires this print to mimic the error from the list-all branch
+      printf "No compatible versions available (%s %s)\n" "$plugin_name" "$query" >&2
+      exit 1
+    fi
+  else
+    # pattern from xxenv-latest (https://github.com/momo-lab/xxenv-latest)
+    versions=$(asdf list-all "$plugin_name" "$query" |
+      grep -vE "(^Available versions:|-src|-dev|-latest|-stm|[-\\.]rc|-alpha|-beta|[-\\.]pre|-next|(a|b|c)[0-9]+|snapshot|master)" |
```
Comment (Stratus3D) [inline lib/commands/command-latest.bash:27]: I know this isn't a new change, but I think `Available versions` could be left out of this list. Plugins should just be printing the list of available versions and nothing else. An `Available versions` heading is not what is specified in the docs.

**Comment by Stratus3D** [comment]
@jthegedus thank you for handling. I am sorry I dropped the ball on this.

@threkk thank you for this PR! I'd been meaning to implement this on my own for months now, this is a great improvement!

File: lib/commands/command-latest.bash (hunk 1 above)
Comment (threkk) [reply lib/commands/command-latest.bash:18]: Sure! It sounds good. I don't mind to look into it when the time comes.

File: lib/commands/command-latest.bash (hunk 2 above)
Comment (threkk) [reply lib/commands/command-latest.bash:27]: Makes sense. I decided to leave it as it was purely for backward compatibility, just in case someone relied on it.
